File: engine/api/src/main/java/io/deephaven/engine/table/hierarchical/TreeTable.java
//
// Copyright (c) 2016-2024 Deephaven Data Labs and Patent Pending
//
package io.deephaven.engine.table.hierarchical;

import io.deephaven.api.*;
import io.deephaven.api.filter.Filter;
import io.deephaven.engine.liveness.LivenessScopeStack;
import io.deephaven.engine.table.ColumnDefinition;
import io.deephaven.engine.table.Table;
import io.deephaven.engine.table.TableDefinition;
import io.deephaven.engine.updategraph.DynamicNode;
import org.jetbrains.annotations.NotNull;

import java.util.Collection;
import java.util.List;

/**
 * Interface for the result of {@link Table#tree(String, String)} tree} operations.
 */
public interface TreeTable extends HierarchicalTable<TreeTable> {

    /**
     * Get the identifier column from the {@link Table#tree(String, String) tree} operation.
     *
     * @return The identifier column
     */
    ColumnName getIdentifierColumn();

    /**
     * Get the parent identifier column from the {{@link Table#tree(String, String) tree} operation.
     *
     * @return The parent identifier column
     */
    ColumnName getParentIdentifierColumn();

    /**
     * Get the {@link TableDefinition} that should be exposed to node table consumers, e.g. UI-driven snapshots. This
     * excludes "internal" columns used to organize the tree or support operations, as well as the
     * {@link #getRowDepthColumn() row-depth column} and {@link #getRowExpandedColumn() row-expanded column}, but
     * includes formatting columns.
     *
     * @return The externally-visible node {@link TableDefinition}
     */
    TableDefinition getNodeDefinition();

    /**
     * Get a new TreeTable with {@code columns} designated for node-level filtering, in addition to any columns already
     * so-designated on {@code this} TreeTable.
     * <p>
     * Filters specified via {@link #withFilter(Filter)}, typically from the UI, that only use the designated node-level
     * filtering columns will be applied to the nodes during snapshots. If no node-filter columns are designated, no
     * filters will be handled at node level.
     * <p>
     * Filters that include other columns are handled by filtering the source table in a ancestor-preserving manner and
     * re-applying the tree operation to the result to produce a new TreeTable. Users of orphan promotion or other
     * strategies to govern the structure of the tree should carefully consider the structure of their data before
     * specifying node-filter columns.
     * <p>
     * Specifying node-filter columns represents a trade-off between performance (which is expected to be much better
     * for node-level filtering) and tree structural integrity (which may be lost since node-level filters are not
     * ancestor-preserving).
     *
     * @param columns The columns to designate
     * @return The new TreeTable
     */
    TreeTable withNodeFilterColumns(@NotNull Collection<? extends ColumnName> columns);

    /**
     * Apply a filter to the columns of this TreeTable in order to produce a new TreeTable.
     *
     * @param filter The filter to apply
     * @return The new TreeTable
     */
    TreeTable withFilter(@NotNull Filter filter);

    /**
     * Recorder for node-level operations to be applied when gathering snapshots.
     */
    interface NodeOperationsRecorder extends
            FormatOperationsRecorder<NodeOperationsRecorder>,
            SortOperationsRecorder<NodeOperationsRecorder>,
            FilterOperationsRecorder<NodeOperationsRecorder> {

        /**
         * Test if this NodeOperationsRecorder has recorded any operations.
         *
         * @return Whether this NodeOperationsRecorder has recorded any operations
         */
        boolean isEmpty();
    }

    /**
     * Get a {@link NodeOperationsRecorder recorder} for per-node operations to apply during snapshots.
     *
     * @return The new recorder
     */
    NodeOperationsRecorder makeNodeOperationsRecorder();

    /**
     * Get a new TreeTable that will apply the {@link NodeOperationsRecorder recorded} operations to nodes when
     * gathering snapshots.
     *
     * @param nodeOperations The node-level operations to apply. Must have been initially supplied by
     *        {@link #makeNodeOperationsRecorder()} from {@code this} TreeTable.
     * @return The new TreeTable
     */
    TreeTable withNodeOperations(@NotNull NodeOperationsRecorder nodeOperations);

    /**
     * Adapt a {@code source} {@link Table} to be used for a {@link Table#tree(String, String) tree} to ensure that the
     * result will have no orphaned nodes. Nodes whose parents do not exist will become children of the root node in the
     * resulting tree. The expected usage pattern is:
     * 
     * <pre>
     * TreeTable result = promoteOrphans(source, idColumn, parentColumn).tree(idColumn, parentColumn)
     * </pre>
     * 
     * @param source The source {@link Table}
     * @param idColumn The name of a column containing a unique identifier for a particular row in the table
     * @param parentColumn The name of a column containing the parent's identifier, {@code null} for rows that are part
     *        of the root table
     * @return A {@link Table} derived from {@code source} that has {@code null} as the parent for any nodes that would
     *         have been orphaned by a call to {@code source.tree(idColumn, parentColumn)}
     */
    static Table promoteOrphans(
            @NotNull final Table source,
            @NotNull final String idColumn,
            @NotNull final String parentColumn) {
        final ColumnName parent = ColumnName.of(parentColumn);
        final ColumnName identifier = ColumnName.of(idColumn);
        final ColumnName sentinel = ColumnName.of("__MATCHED_PARENT_IDENTIFIER__");
        final Selectable[] viewColumns = source.getDefinition().getColumnStream()
                .map(ColumnDefinition::getName)
                .map((final String columnName) -> columnName.equals(parent.name())
                        ? Selectable.of(parent,
                                RawString.of("isNull(" + sentinel.name() + ") ? null : " + parent.name()))
                        : ColumnName.of(columnName))
                .toArray(Selectable[]::new);
        return LivenessScopeStack.computeEnclosed(
                () -> source
                        .naturalJoin(source,
                                List.of(JoinMatch.of(parent, identifier)),
                                List.of(JoinAddition.of(sentinel, identifier)))
                        .view(List.of(viewColumns)),
                source::isRefreshing,
                DynamicNode::isRefreshing);
    }
}

File: py/server/deephaven/tree_table.py
#
#     Copyright (c) 2016-2022 Deephaven Data Labs and Patent Pending
#
"""This module defines the TreeTable and the operations that can be performed on instances of TreeTable."""

from __future__ import annotations

from typing import Sequence, Union

import jpy

from deephaven import DHError
from deephaven._wrapper import JObjectWrapper
from deephaven.filters import Filter
from deephaven.jcompat import to_sequence, j_array_list
from deephaven.rollup_table import FormatOperationsRecorder, SortOperationsRecorder, FilterOperationsRecorder

_JTreeTable = jpy.get_type("io.deephaven.engine.table.hierarchical.TreeTable")
_JNodeOperationsRecorder = jpy.get_type("io.deephaven.engine.table.hierarchical.TreeTable$NodeOperationsRecorder")


class TreeNodeOperationsRecorder(JObjectWrapper, FormatOperationsRecorder,
                                 SortOperationsRecorder, FilterOperationsRecorder):
    """Recorder for node-level operations to be applied when gathering snapshots of TreeTable. Supported operations
    include column formatting, sorting, and filtering.

    Note: It should not be instantiated directly. User code must call :meth:`~TreeTable.node_operation_recorder` to
    create an instance of the recorder.
    """

    j_object_type = _JNodeOperationsRecorder

    @property
    def j_object(self) -> jpy.JType:
        return self.j_node_ops_recorder

    def __init__(self, j_node_ops_recorder: jpy.JType):
        self.j_node_ops_recorder = j_node_ops_recorder


class TreeTable(JObjectWrapper):
    """ A TreeTable is generated as a result of applying the :meth:`~Table.tree` method on a
    :class:`~deephaven.table.Table`.

    Note: TreeTable should not be instantiated directly by user code.
    """
    j_object_type = _JTreeTable

    @property
    def j_object(self) -> jpy.JType:
        return self.j_tree_table

    def __init__(self, j_tree_table: jpy.JType, id_col: str, parent_col: str):
        self.j_tree_table = j_tree_table
        self.id_col = id_col
        self.parent_col = parent_col

    def node_operation_recorder(self) -> TreeNodeOperationsRecorder:
        """Creates a TreepNodeOperationsRecorder for per-node operations to apply during snapshotting of this TreeTable.

        Returns:
            a TreeNodeOperationsRecorder
        """
        return TreeNodeOperationsRecorder(j_node_ops_recorder=self.j_tree_table.makeNodeOperationsRecorder())

    def with_node_operations(self, node_ops: TreeNodeOperationsRecorder) -> TreeTable:
        """Returns a new TreeTable that will apply the recorded node operations to nodes when gathering snapshots.

        Args:
            node_ops (TreeNodeOperationsRecorder): the TreeNodeOperationsRecorder containing the node operations to be
                applied

        Returns:
            a new TreeTable

        Raises:
            DHError
        """

        try:
            return TreeTable(
                j_tree_table=self.j_tree_table.withNodeOperations(node_ops.j_node_ops_recorder),
                id_col=self.id_col, parent_col=self.parent_col)
        except Exception as e:
            raise DHError(e, "with_node_operations on TreeTable failed.") from e

    def with_filters(self, filters: Union[str, Filter, Sequence[str], Sequence[Filter]]) -> TreeTable:
        """Applies a set of filters to the columns of this TreeTable to produce and return a new TreeTable

        Args:
            filters (Union[str, Filter, Sequence[str], Sequence[Filter]], optional): the filter condition
                expression(s) or Filter object(s)

        Returns:
            a new TreeTable

        Raises:
            DHError
        """

        try:
            filters = to_sequence(filters)
            if filters and isinstance(filters[0], str):
                filters = Filter.from_(filters)
            filters = j_array_list(filters)

            return TreeTable(j_tree_table=self.j_tree_table.withFilters(filters), id_col=self.id_col,
                             parent_col=self.parent_col)
        except Exception as e:
            raise DHError(e, "with_filters operation on TreeTable failed.") from e
